grr/server/grr_response_server/gui/api_plugins/metadata.py

```diff
@@ -268,9 +263,82 @@ def _CreateMessageSchema(
       }
     )
 
+  def _CreateMapFieldSchema(
+      self,
+      field_descriptor: FieldDescriptor,
+      visiting: Set[str],
+  ) -> None:
+    """Creates the OpenAPI schema of a `protobuf.map` field type.
+
+    This method should generally not be called directly, but rather through
+    `_CreateSchema` which takes care of error-verifications and caching. The
+    OpenAPI Specification allows only maps with strings as keys, so, as a
+    workaround, we state the actual key type in the `description` fields of the
+    **schemas of the properties/parameters** that use this type (in order to be
+    displayed by documentation generation tools, see `_GetDescribedSchema`).
+    A `description` field is also added by this method in the schema definition
+    which will be added to the `Components Object` of the root `OpenAPI Object`
+    for more clarity when reading the generated description of the components.
+
+    Args:
+      field_descriptor: The protobuf `FieldDescriptor` associated with a
+        protobuf field.
+      visiting: A set of type names that are in the process of having their
+        OpenAPI schemas constructed and have their associated `_Create*Schema`
+        call in the current call stack.
+
+    Returns:
+      Nothing, the schema is stored in `self.schema_objs`.
+    """
+    if self.schema_objs is None:  # Check required by mypy.
+      raise AssertionError(
+        "The container of OpenAPI type schemas is not initialized."
+      )
+
+    if field_descriptor is None:  # Check required by mypy.
+      raise AssertionError(f"`field_descriptor` is None.")
+
+    type_name: str = _GetTypeName(field_descriptor)
+    visiting.add(type_name)
+
+    key_field_d, value_field_d = (
+      _GetMapFieldKeyValueTypes(field_descriptor)
+    )
+
+    # pylint: disable=line-too-long
+    # `protobuf.map` key types can be only a subset of the primitive types [1],
+    # so there is definitely no composite key type to further visit, but the
+    # value type "can be any type except another map" [1] or an array [2].
+    #
+    # [1]: https://developers.google.com/protocol-buffers/docs/proto#maps
+    # [2]: https://developers.google.com/protocol-buffers/docs/reference/proto2-spec#map_field
+    # pylint: enable=line-too-long
+    self._CreateSchema(value_field_d, visiting)
+
+    if key_field_d is None:  # Check required by mypy.
+      raise AssertionError(
+        f"The map entry message type has no `key` FieldDescriptor."
+      )
+
+    visiting.remove(type_name)
+
+    self.schema_objs[type_name] = cast(
+      Dict[str, Union[str, SchemaReference]],
+      {
+        "description": f"This is a map with real key type="
+                       f"\"{_GetTypeName(key_field_d)}\" and value type="
+                       f"\"{_GetTypeName(value_field_d)}\"",
+        "type": "object",
+        "additionalProperties": _GetReferenceObject(
+          _GetTypeName(value_field_d)),
```

Comment:
I would move this to a separate variable, it's hard to parse nested method calls with a line break (for a moment I thought that this is a pair).

Reply:
Thank you for your suggestion! I created variables for the key and value type names, this removes the nested method calls and also avoids duplication with the code above. Please let me know if you would like me to treat it differently (such as creating a separate variable only/also for the `Reference Object`).